### PR TITLE
Restart: add restart button to restart emulation easier

### PR DESCRIPTION
**Pull request by JasonMo1**

### CODE REVIEW

File: machine.js [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+            return obj;
+        },
+
+        removeTstateCallback: function(callback) {
+            if (callback != null) {
+                tstates_callbacks.delete(callback);
+            }
+        },
+
+        interrupt: function(interrupt_vector) {
```
Comment (Zeal8bit) [inline machine.js:110]: This function is not related to T-States, it should be taken out

File: machine.js [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+                return;
+            }
+            /* Here, the number of T-state the callback is meant to be executed is in the range
+            * [t_state;end], so it is meant to happen during this iteration.
+            * Jump to that amount and execute instructions following it directly. */
+            t_state = earliest.tstates;
+        }
+    }
+
+    this.tstatesutils = tstatesutils;
```
Comment (Zeal8bit) [inline machine.js:129]: I like the fact that all the functions related to T-States callbacks have been gathered together. Can you keep this object private and create a public API for it? Like:
```
this.getTstates = tstatesutils.getTstates;
this.registerTstatesCallback = tstatesutils.registerTstateCallback;
// etc...
```

File: machine.js [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    this.uart = uart;
+    this.i2c = i2c;
+    this.keyboard = keyboard;
+    this.ds1307 = ds1307;
+    this.eeprom = eeprom;
+    this.devices = devices;
+    this.zpu = zpu;
+
+    var registers = zpu.getState();
+    this.registers = registers;
```
Comment (Zeal8bit) [inline machine.js:163]: Similarly for this, can we make it private and create a public function for retrieving the structure?

File: zeal.js [hunk 4]
```diff
@@ -479,25 +194,24 @@ $("#read-button").on('click', function() {
...
+    const handled = zealcom.keyboard.key_pressed(e.keyCode);
 
     if (handled) {
         e.preventDefault();
     }
 });
 
 $("#screen").on("keyup", function(e) {
-    const handled = keyboard.key_released(e.keyCode);
+    const handled = zealcom.keyboard.key_released(e.keyCode);
```
Comment (Zeal8bit) [inline zeal.js:214]: Similarly, I wonder if it would be more meaningful to have a public API for this:
```
KeyboardKeyPressed(e.keyCode);
KeyboardKeyRelease(e.keyCode);
```
given by the `Machine` object

**Review by Zeal8bit — COMMENTED** [review]
Thanks for your contribution! I have a few comments but it looks great already!

File: machine.js (hunk 1 above)
Comment (JasonMo1) [reply machine.js]: Fixed.

File: machine.js (hunk 2 above)
Comment (JasonMo1) [reply machine.js]: Fixed.

File: machine.js (hunk 3 above)
Comment (JasonMo1) [reply machine.js]: Fixed.

File: zeal.js (hunk 4 above)
Comment (JasonMo1) [reply zeal.js]: Added.

**Comment by Zeal8bit** [comment]
Thanks for your contribution, merging!
